source/agora/consensus/EnrollmentManager.d

```diff
@@ -35,6 +35,19 @@ import std.path;
 
 mixin AddLogger!();
 
+/*******************************************************************************
+
+    Define preimage data for a preimage at a certain round and the round number
+
+*******************************************************************************/
+
+public struct PreimageInfo
+{
+    public Hash utxo_key;
+    public Hash hash;
+    public uint round;
```

Comment:
That's an interesting approach. When we were originally talking about it, I had in mind that this would by *block height*, not round. Because the issue is that if you re-use the same `utxo_key` (as you would), then you can do replay attacks.
Of course validation would fail, but you could potentially pretend a node sent you the wrong preimage data. We didn't factor any slashing rules that could be triggered by this tho.

A downside of using the `block_height` is that it's harder to pre-emptively broadcast preimage as you don't know whe height at which it'll be included. But that can be dealt with when the enrollment is received.

Reply:
In any case, this could use some documentation

Reply:
I'd like accept your thought because it's more convenient to recognize the block to which the preimage is related. So, I could add the variable named `start_block_height` into the Enrollment structure which is the block height where a validator is supposed to participate in validating. Then, The variable, `round` of the PreimageInfo structure could be replaced with a variable like `round_block_height`. How about this?

Reply:
> start_block_height

Minor stylistic note: "height" is pretty unambiguous in blockchain code, so you can just use `start_height` (because there's no other height that block height).

> which is the block height where a validator is supposed to participate in validating

I'll have to double check, the exact flow, but bear in mind that if we go this way, it is an *implicit* value. What I mean is that, when you create an `Enrollment` structure and sign it, you cannot know which `start_height` you will be assigned, because it has to be accepted by the validators. Same goes for transactions, when you create a transaction, you cannot know at which height it will be.

Does that make sense ? Also, I think both approaches have their merits, so maybe let's have a whiteboarding session about it ?

Reply:
I understand what you're considering. After a bit of thought, I can use the `enrolled_height` variable which is implemented by Henry. The value of `enrolled_height + 1` will have the same value as what I want to have with the `start_block_height` variable. It's good to have a whiteboarding session.